python/analyze.py

```diff
@@ -5,46 +5,121 @@
-        for i in nffts:
-            if i <= audio.size:
-                nfft = i
-        nffts = 256
+        audio, srate = librosa.load(dir, sr = sr, mono = True, duration = duration, res_type='kaiser_fast')
+        audio, index = librosa.effects.trim(y = audio, top_db=60, frame_length=2048, hop_length=512)
+        nfft = 2048
+        hop = 1024
+
+        # 0 pad to 2048 minimum size
+        if audio.size < 2048:
+            audio = np.pad(audio, (0, 2048 - audio.size), constant_values=(0, 0))
+
+
         # Max size limits the length of audio sampls so that all features will have the same length (0 padding)
-        max_size = min(88200, audio.size)
-        mfccs = librosa.feature.mfcc(y=audio[:max_size], sr=srate, n_mfcc=40, n_fft=nfft)
-        pad_width = 174 - mfccs.shape[1]
-        mfccs = np.pad(mfccs, pad_width=((0,0), (0, pad_width)), mode='constant')
-        #mfcc = np.mean(mfccs.T, axis=0)
-        return mfccs
+        max_size = min(srate * 3, audio.size)
+        mfccs = librosa.feature.mfcc(y=audio, sr=srate, n_mfcc=40, n_fft=nfft, hop_length = hop, power = 2.0)
+
+        pad_width = 108 - mfccs.shape[1]
+        mfccs = np.pad(mfccs, pad_width=((0,0), (0, pad_width)), mode='reflect')
+        
+
+        centroid = librosa.feature.spectral_centroid(y=audio, sr = srate, n_fft = nfft, hop_length = hop)
+        bandwidth = librosa.feature.spectral_centroid(y= audio, sr = srate, n_fft = nfft, hop_length = hop)
+        spectral_contrast = librosa.feature.spectral_contrast(y = audio, sr = srate, n_fft = nfft, hop_length = hop)
+        spectral_flatness = librosa.feature.spectral_flatness(y = audio, n_fft = nfft, hop_length = hop)
+        spectral_rolloff = librosa.feature.spectral_rolloff(y = audio, sr = srate, n_fft = nfft, hop_length = hop)
+        zcr = librosa.feature.zero_crossing_rate(y = audio, hop_length = 1024)
+
+
+        feature_vector = [
+            centroid.mean(),
+            centroid.std(),
+            bandwidth.mean(),
+            bandwidth.std(),
+            # spectral_contrast[0].mean(),
+            # spectral_contrast[0].std(),
+            # spectral_contrast[1].mean(),
+            # spectral_contrast[1].std(),
+            # spectral_contrast[2].mean(),
+            # spectral_contrast[2].std(),
+            # spectral_contrast[3].mean(),
+            # spectral_contrast[3].std(),
+            # spectral_contrast[4].mean(),
+            # spectral_contrast[4].std(),
+            # spectral_contrast[5].mean(),
+            # spectral_contrast[5].std(),
+            spectral_flatness.mean(),
+            spectral_flatness.std(),
+            spectral_rolloff.mean(),
+            spectral_rolloff.std(),
+            zcr.mean(),
+            zcr.std(),
+            mfccs.mean(),
+            mfccs.std()
+        ]
+        
+        feature_vector = np.array(feature_vector)
+        normalized_feature_vector = (feature_vector - feature_vector.mean(0)) / feature_vector.std(0)
```

Comment:
Is that the mean and std of the feature vector itself? How does this work when each feature may have a different range and units?

Reply:
If the feature vector represents the entire dir then that makes sense.